src/arcade/patch/agent/cell/PatchCell.java

```diff
@@ -439,92 +442,127 @@ public static double calculateTotalVolume(Bag bag) {
-                if (currentHeight > targetHeight) {
-                    continue;
-                }
+    /**
+     * Find free locations in the neighborhood of the cell.
+     *
+     * @param sim the simulation instance
+     * @param add true if an additional cell is being added
+     * @return a list of free locations
+     */
+    public Bag findFreeLocations(Simulation sim, boolean add) {
+        Bag freeLocations = new Bag();
 
-                // Check if neighbor cells can exist at a tolerable height.
-                for (Object obj : bag) {
-                    if (currentHeight > ((Cell) obj).getCriticalHeight()) {
-                        continue locationCheck;
-                    }
-                }
+        PatchLocation currentLocation = this.location;
+        double targetVolume;
+        if (add) {
+            targetVolume = this.getVolume() * 0.5;
+        } else {
+            targetVolume = this.getVolume();
+        }
 
-                // TODO: ADD CHECK FOR MORE THAN ONE HEALTHY CELL AGENT.
+        if (checkLocation(
+                sim, currentLocation, 0, getCriticalHeight(), getPop(), MAX_DENSITY, add)) {
+            freeLocations.add(currentLocation);
+        }
 
-                // Add location to list of free locations.
+        for (Location neighborLocation : currentLocation.getNeighbors()) {
+            PatchLocation neighbor = (PatchLocation) neighborLocation;
+
+            if (checkLocation(
+                    sim,
+                    neighbor,
+                    targetVolume,
+                    getCriticalHeight(),
+                    getPop(),
+                    MAX_DENSITY,
+                    true)) {
                 freeLocations.add(neighborLocation);
             }
         }
-
-        // TODO: ADD CURRENT LOCATION
-
         return freeLocations;
     }
 
     /**
-     * Selects the best location for a cell to be added or move into.
-     *
-     * <p>Each free location is scored based on glucose availability and distance from the center of
-     * the simulation.
+     * Determine if a patch location is free.
      *
      * @param sim the simulation instance
-     * @param location the current location
-     * @param volume the target volume of cell to add or move
-     * @param height the target height of the cell to add or move
-     * @param random the random number generator
-     * @return the best location
+     * @param loc the location
+     * @param targetVolume the target volume of the cell to add or move
+     * @param targetHeight the target height of the cell to add or move
+     * @param population the population index
+     * @param maxDensity the maximum density of population in the location
+     * @param add true if an additional cell is being added
+     * @return a list of free locations
      */
-    public static PatchLocation selectBestLocation(
+    static boolean checkLocation(
```

Comment:
I am a little confused by the purpose of the `add` parameter:

When would there be a case where the target volume is non-zero, but the add boolean is false?  When would checkLocation be called that doesn't involve another cell being potentially added into the location? Even if the cell is moving instead of proliferating, doesn't the number of cells in the receiving location still increase by 1?

Reply:
So there's a couple cases for the current location to consider: 
- A migrating cell should still be able to return that the current location is a free location, unless there are other constraints (increase in height) that are preventing it from staying in this location. The availability of the current location shouldn't be impacted by there being an additional cell in this case. This case would be a target volume of 0 and add would be false. 
- A dividing cell is not necessarily adding volume to the current location, but it is increasing the number of constituents, so a location that a cell can divide into is dependent on being able to hold an additional cell rather than an increase in volume. This case would be a target volume of 0 and add would be true. 
- I don't think there's a condition where the target volume would be non-zero and add would be false. 

I'm not sure if I have a better solution for these cases? Would wrapping them in clearer function names make the arguments more intuitive?

Reply:
I agree the `add`/`proliferative` booleans are a little confusing. I haven't fully thought this through, but some options:

- simplify max density by changing it to a boolean `allowMultiple` (or something) where it is `true` for cancer cells but `false` for healthy cells (pass this instead of `maxDensity` and `add` to `checkLocation`)
- pass the `includeCurrent` boolean to `findFreeLocations` and just skip the current location check entirely if it is false
- rename the two `selectBestLocation` methods to something like `selectBestLocationToAdd` and `selectBestLocationToMove`?

Reply:
I am trying to think through if we need to have the max cell density be an actual integer or just a boolean:

Are there cases in which the limit is actually somewhere between 1 and infinity? For T cells, CARCADE has the limit as 54, but I am not sure if there is a potential for T cells to actually reach this number or if is again just a stand-in for inf. 

I agree with Jessica's second and third bullet point though

Reply:
I've changed the implementation flag such that we can handle integer max cell densities. There are likely cases where you might want to limit computational complexity or enforce biological density (similar to healthy cells), but is now a user option. 

As for the proliferation flag, I've removed it from the check location and the flag updates the arguments to the static method.